src/transaction/atomic_replication_helper.cpp

```diff
@@ -301,162 +354,410 @@ namespace cublog
     : m_vpid { vpid }
+    , m_rectype { LOG_LARGER_LOGREC_TYPE }
     , m_record_lsa { lsa }
     , m_record_index { rcvindex }
+    , m_sysop_end_type { (LOG_SYSOP_END_TYPE)-1 }
+    , m_sysop_end_last_parent_lsa { NULL_LSA }
     , m_page_ptr { page_ptr }
   {
     assert (!VPID_ISNULL (&m_vpid));
     assert (m_record_lsa != NULL_LSA);
-    assert (0 <= m_record_index && m_record_index <= RV_LAST_LOGID);
+    assert (0 <= m_record_index &&m_record_index <= RV_LAST_LOGID);
     assert (m_page_ptr != nullptr);
   }
 
+  atomic_replication_helper::atomic_log_sequence::atomic_log_entry::atomic_log_entry (
+	  LOG_LSA lsa, LOG_RECTYPE rectype)
+    : m_vpid VPID_INITIALIZER
+    , m_rectype { rectype }
+    , m_record_lsa { lsa }
+    , m_record_index { RV_NOT_DEFINED }
+    , m_sysop_end_type { (LOG_SYSOP_END_TYPE)-1 }
+    , m_sysop_end_last_parent_lsa { NULL_LSA }
+    , m_page_ptr { nullptr }
+  {
+    assert (m_record_lsa != NULL_LSA);
+    assert (m_rectype != LOG_LARGER_LOGREC_TYPE );
+    // there is a specific ctor for sysop end
+    assert (m_rectype != LOG_SYSOP_END );
+  }
+
+  atomic_replication_helper::atomic_log_sequence::atomic_log_entry::atomic_log_entry (
+	  LOG_LSA lsa, LOG_SYSOP_END_TYPE sysop_end_type, LOG_LSA sysop_end_last_parent_lsa)
+    : m_vpid VPID_INITIALIZER
+    , m_rectype { LOG_SYSOP_END }
+    , m_record_lsa { lsa }
+    , m_record_index { RV_NOT_DEFINED }
+    , m_sysop_end_type { sysop_end_type }
+    , m_sysop_end_last_parent_lsa { sysop_end_last_parent_lsa }
+    , m_page_ptr { nullptr }
+  {
+    assert (m_record_lsa != NULL_LSA);
+    assert (LOG_SYSOP_END_COMMIT <= sysop_end_type &&sysop_end_type <= LOG_SYSOP_END_LOGICAL_RUN_POSTPONE);
+    // sysop end parent lsa can also be null
+  }
+
   atomic_replication_helper::atomic_log_sequence::atomic_log_entry::atomic_log_entry (atomic_log_entry &&that)
-    : atomic_log_entry (that.m_record_lsa, that.m_vpid, that.m_record_index, that.m_page_ptr)
+    : m_vpid { that.m_vpid }
+    , m_rectype { that.m_rectype }
+    , m_record_lsa { that.m_record_lsa }
+    , m_record_index { that.m_record_index }
+    , m_sysop_end_type { that.m_sysop_end_type }
+    , m_sysop_end_last_parent_lsa { that.m_sysop_end_last_parent_lsa }
+    , m_page_ptr { that.m_page_ptr }
   {
   }
 
+  atomic_replication_helper::atomic_log_sequence::atomic_log_entry &
+  atomic_replication_helper::atomic_log_sequence::atomic_log_entry::operator= (atomic_log_entry &&that)
+  {
+    std::swap (m_vpid, that.m_vpid);
+    std::swap (m_rectype, that.m_rectype);
+    std::swap (m_record_lsa, that.m_record_lsa);
+    std::swap (m_record_index, that.m_record_index);
+    std::swap (m_sysop_end_type, that.m_sysop_end_type);
+    std::swap (m_sysop_end_last_parent_lsa, that.m_sysop_end_last_parent_lsa);
+    std::swap (m_page_ptr, that.m_page_ptr);
+    return *this;
+  }
+
   void
   atomic_replication_helper::atomic_log_sequence::atomic_log_entry::apply_log_redo (
```

Comment:
Could you tell me why only a few record types are handled here? Isn't it true that any record types can be put in an atomic system operation such as LOG_MVCC_DIFF_UNDOREDO_DATA?

Reply:
These are the ones that have been encountered in tests. Notice the
```
      default:
	assert (false);
	break;
```
at the end.

Reply:
I think this is making a restriction on atomic system operation. The way I see it, system operations are designed to be able to contain any kind of operation. I'm quite against that we can ignore their usable cases just because they haven't been used in that way. 
 If it is necessary to make that restriction, we should not allow users to make a sysop with other log records. Or, do you have a plan to extend this to cover others later?

Reply:
Of course these can be extended later. It will be the job off different test scenarios to uncover uncovered cases.
I don't see any restriction.

However, and I now see this, The `LOG_RECTYPE`'s that are handled in `atomic_log_entry::apply_log_redo` are the same that are handled via `atomic_replicator::redo_upto` -> `atomic_replicator::read_and_redo_record` -> `atomic_replication_helper::append_log` -> `atomic_log_sequence::append_log`.

And these switches have been taken over from `log_recovery_redo` which was our "single source of truth" when we began replication.

> I'm quite against that we can ignore their usable cases just because they haven't been used in that way.
I don't think I understand.

> Do you have a plan to extend this to cover others later?
Whenever needed.

Reply:
> And these switches have been taken over from log_recovery_redo which was our "single source of truth" when we began replication.

Oh,, then I'm happy with it. I thought it does take over only some parts of that. An example I mentioned, LOG_MVCC_DIFF_UNDOREDO_DATA, is handled, but I missed it. Sorry to bother you with it.

Reply:
They only case omitted seems to be LOG_DBEXTERN_REDO_DATA. Is it on purpose?

Reply:
`LOG_DBEXTERN_REDO_DATA` is handled directly in `atomic_replicator::redo_upto`. It never gets into an atomic sequence. Because it does not change a certain page.
`LOG_DBEXTERN_REDO_DATA` seems to have only 3 rcvindices (as investigated via `log_append_dboutside_redo` -> `prior_lsa_alloc_and_copy_data` -> `prior_lsa_gen_dbout_redo_record`):
- RVDK_NEWVOL, RVDK_EXPAND_VOLUME - both skipped for atomic replication
- `RVLOG_OUTSIDE_LOGICAL_REDO_NOOP`  whose redo function is no-op